finagle-http/src/main/scala/com/twitter/finagle/http/Codec.scala

```diff
@@ -521,3 +522,95 @@ private[finagle] class HttpServerTracingFilter[Req <: Request, Res](serviceName:
     service(request)
   }
 }
+
+private[finagle] object OpenTracing {
+
+  import io.opentracing.Tracer
+  import io.opentracing.Span
+  import java.util.HashMap
+  import scala.collection.JavaConverters._
+
+  private[this] val REQUEST = "Finagle HTTP Request"
+  private[this] val RESPONSE = "Finagle HTTP Response"
+
+  /**
+   * @param span the parent span of all requests created during the 
+   *  scope of the enclosed function.
+   */
+  def letCurrentSpan[R](span: Span)(f: => R): R = 
+    Contexts.local.let(spanCtx, span)(f)
+
+  private[this] val spanCtx = Contexts.local.newKey[Object]
+
+  /**
+   * Creates a new span for the HTTP request and injects the necessary 
+   * parameters so the server can join the trace.
+   * 
+   * The parent span of the request can be set by calling letCurrentSpan
+   * on the request.
+   * @param tracer the OpenTracing tracer to create the span
+   */
+  class ClientFilter[Req <: Request, Res](tracer: Tracer) 
+    extends SimpleFilter[Req, Res]
+  {
+    def apply(request: Req, service: Service[Req, Res]) = {
+      val span = createSpanFromContext(tracer, REQUEST)
+
+      // add parameters necessary to recreate this span across 
+      // the wire to the HTTP headers
+      var textMap = new HashMap[String,String]()
+      tracer.inject(span, textMap)
+      for((k:String, v:String) <- textMap.asScala) { request.headers.add(k, v) }
+
+      // close the span once the response is received
+      val response = service(request).respond { resp =>
+        span.close()
+      } 
+      response
+    }
+  }
+
+  /**
+   * Traces this HTTP response using the injected headers. Depending on the 
+   * OpenTracing implementation, this may continue the span from the 
+   * client request or create a new child span for the server response.
+   * If there is no injected span to join, creates a new span from the 
+   * current span context, which can be set using letCurrentSpan.
+   */
+  class ServerFilter[Req <: Request, Res](tracer: Tracer)
+    extends SimpleFilter[Req, Res]
+  {
+    def apply(request: Req, service: Service[Req, Res]) = {
+      // extract the span tracking the client request
+      var textMap: HashMap[String, String] = new HashMap[String, String]()
+      for((k: String, v:String) <- request.headerMap) {
+        textMap.put(k,v)
+      }
+      var span: Span = null
+      try {
+        span = tracer.join(textMap).withOperationName(RESPONSE).start()
+      } 
+      // If the join fails because of corrupt headers or no client-side 
+      // filter, a new span is created and the failure is noted.
+      catch {
+        case _: Throwable => {
+          span = createSpanFromContext(tracer, RESPONSE)
+          span.log("Join failed", null)
+        }
+      }
+      val response = service(request)
+      if(span != null){ span.close() }
```

Comment:
probably better to use that `ensure` statement if you can figure out how it works :)